exonum/src/node/state.rs

```diff
@@ -851,6 +910,36 @@ impl State {
         }
     }
 
+    /// Finds unknown transactions in the block.
+    pub fn create_incomplete_block(
+        &mut self,
+        msg: &BlockResponse,
+        txs: &MapIndex<&&Snapshot, Hash, RawMessage>,
+        txs_pool: &KeySetIndex<&&Snapshot, Hash>,
+    ) -> Result<bool, failure::Error> {
+        let mut unknown_txs = HashSet::new();
+        for hash in msg.transactions() {
+            if txs.get(hash).is_some() {
+                if !txs_pool.contains(hash) {
+                    bail!(
+                        "Received block with already \
+                         committed transaction"
+                    )
+                }
+            } else {
+                unknown_txs.insert(*hash);
+            }
+        }
+
+        let response = Ok(!unknown_txs.is_empty());
+        self.incomplete_block = Some(IncompleteBlock {
+            msg: msg.clone(),
+            unknown_txs,
+        });
+
+        response
```

Comment:
Nit: can be expressed more succinctly with `self.incomplete_block.has_unknown_txs()`. Alternatively, you may return `&IncompleteBlock` instead of `bool`; IMO, it would be more understandable semantically.